runtime/lua/vim/filetype.lua

```diff
@@ -2213,6 +2186,19 @@ local pattern = {
   end,
   -- END PATTERN
 }
+
+-- BEGIN ENV_PATTERN
+local env_pattern = {
+  ['/cabal%.config'] = { 'cabalconfig', 'HOME' },
```

Comment:
Hmm, maybe we can use a placeholder instead? Something like
```
['$HOME/git/ignore'] = 'gitignore'
```
and then in the helper function, replace it `$HOME` by `vim.env['$HOME']` in the key?

(Maybe not worth it, but would be neat, and maybe allow us to support patterns involving variables elsewhere in the pattern.)

Reply:
That definitely looks nicer, I'll try that.

Reply:
Could be `$HOME$`, then it's easier to split?

Reply:
I like this idea too. Couldn’t we just do this in the existing `pattern` table rather than creating a separate `env_pattern` table?

Reply:
Hmm, I worry a bit about performance implications if we have to check that for every pattern. And I kina like having these special patterns all in one place.

But otherwise no objections!

Reply:
>Hmm, I worry a bit about performance implications if we have to check that for every pattern.

That seems like a reasonable justification to me 👍

Reply:
I've now come up with a solution that avoids the performance implications clason mentioned by making the expansion of environment patterns opt-in while also not requiring a separate `env_patterns` table. What do you think?

Reply:
Nice!

Personally, I think I'd still prefer a separate table over adding a new options mechanism, but either way is OK for me .

And a filename is just a pattern without (or with specific) wildcards, isn't it ;)

Reply:
> And a filename is just a pattern without (or with specific) wildcards, isn't it ;)

That's true ;) Then I will move the `vim.env.HOME` pattern to `patterns`.

Reply:
Another alternative could be something like:
```lua
['.*/cabal%.config'] = expand_env("cabalconfig", "${HOME}/cabal%.config"), 
```
where `expand_env` is a private function. This would avoid adding the new `expand_env` option to the public API.

Reply:
> This would avoid adding the new expand_env option to the public API.

Why do we want to avoid that? It seems like a useful option to provide users in their own filetype.lua files.

Reply:
Yes, that should be available for `filetype.add` as well. The question is only whether we want
1. a separate table for `pattern_env`
2. a separate option for every pattern

Either is fine; 1. seems cleaner, but 2. is more future-proof.